File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if(list1 == null){
            return list2;
        }
        else if(list2 == null){
            return list1;
        }

        ListNode cur1 = list1;
        ListNode cur2 = list2;
        ListNode prev = null;
        ListNode newHead = (cur1.val <= cur2.val) ? cur1 : cur2;

        while(cur1 != null && cur2 != null){
            if(cur1.val <= cur2.val){
                if(prev != null){
                    prev.next = cur1;
                }
                prev = cur1;
                cur1 = cur1.next;
            }
            else{
                if(prev != null){
                    prev.next = cur2;
                }
                prev = cur2;
                cur2 = cur2.next;
            }
        }
        if(cur1 == null){
            prev.next = cur2;
        }
        else if(cur2 == null){
            prev.next = cur1;
        }
        return newHead;
    }
}

File: Data Structures & Algorithms/add-two-numbers/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if l1.val == 0:
            return l2
        elif l2.val == 0:
            return l1

        carryOver = False
        result = ListNode(0)
        resultHead = result

        #while neither number has reached the end
        while l1 and l2:
            val1 = l1.val
            val2 = l2.val
            total = val1 + val2 + carryOver

            if total >= 10:
                carryOver = True
                total -= 10
            else:
                carryOver = False

            result.next = ListNode(total)
            result = result.next
            l1 = l1.next
            l2 = l2.next

        #if l1 & l2 are different in length
        #if l1 still has remaining digits
        while l1:
            cur = l1.val + carryOver
            if carryOver and cur >= 10:
                cur -= 10
            elif cur < 10:
                carryOver = False
            result.next = ListNode(cur)
            result = result.next
            l1 = l1.next
        #if l2 still has remaining digits
        while l2:
            cur = l2.val + carryOver
            if carryOver and cur >= 10:
                cur -= 10
            elif cur < 10:
                carryOver = False
            result.next = ListNode(cur)
            result = result.next
            l2 = l2.next

        #if there's a left over carry
        if carryOver:
            result.next = ListNode(1)

        return resultHead.next
